src/lib.rs

```diff
@@ -1,28 +1,137 @@
+//!    data on a stream. Structures that implement write interfaces include:
+//!    - [`WriteStream`](crate::dataflow::stream::WriteStream): used by
+//!      operators to send data.
+//!    - [`IngestStream`](crate::dataflow::stream::IngestStream): used by
+//!      the driver to send data.
+//!
+//! Some applications may want to introduce loops in their dataflow graphs
+//! which is possible by using the
+//! [`LoopStream`](crate::dataflow::stream::LoopStream) to do so.
+//!
+//! ## Operators
+//! An ERDOS operator receives data on
+//! [`ReadStream`](crate::dataflow::stream::ReadStream)s,
+//! and sends processed data on
+//! [`WriteStream`](crate::dataflow::stream::WriteStream)s.
+//! We provide a [standard library of operators](crate::dataflow::operators)
+//! for common dataflow patterns.
+//! While the standard operators are general and versatile, some applications
+//! may implement custom operators to better optimize performance and take
+//! fine-grained control over exection.
+//!
+//! All operators must implement `new` and `connect` methods, in addition to
+//! the [`Operator`](crate::dataflow::Operator) trait.
+//! - The `new` method takes an
+//!   [`OperatorConfig`](crate::dataflow::OperatorConfig),
+//!   all [`ReadStream`](crate::dataflow::stream::ReadStream)s from which the
+//!   operator receives data, all
+//!   [`WriteStream`](crate::dataflow::stream::WriteStream)s on which the
+//!   operator sends data, and returns `Self`.
+//!   Within `new`, the state should be initialized, and callbacks may be
+//!   registered across [`ReadStream`](crate::dataflow::stream::ReadStream)s.
+//! - The `connect` method takes references to the required
+//!   [`ReadStream`](crate::dataflow::stream::ReadStream)s
+//!   and returns [`WriteStream`](crate::dataflow::stream::WriteStream)s in the
+//!   same order as in `new`.
+//!
+//! For an example, see the implementation of the
+//! [`MapOperator`](crate::dataflow::operators::MapOperator).
+//!
+//! While ERDOS manages the execution of callbacks, some operators require
+//! more finegrained control. Operators can take manual control over the
+//! thread of execution by overriding the
+//! [`run`](crate::dataflow::Operator::run) of the
+//! [`Operator`](crate::dataflow::Operator) trait and pulling data from
+//! [`ReadStream`](crate::dataflow::stream::ReadStream)s.
+//!
+//! ## Performance
 //! `ERDOS` is designed for low latency. Self-driving car pipelines require
 //! end-to-end deadlines on the order of hundreds of milliseconds for safe
 //! driving. Similarly, self-driving cars typically process gigabytes per
 //! second of data on small clusters. Therefore, `ERDOS` is optimized to
 //! send small amounts of data (gigabytes as opposed to terabytes)
 //! as quickly as possible.
 //!
-//! `ERDOS` provides determinisim through **watermarks**. Low watermarks
+//! ## Determinism
+//! ERDOS provides determinisim through **watermarks**. Low watermarks
```

Comment:
Can you please also add a section on watermarks because they're a core feature of ERDOS? I think we also need to explain top & bottom watermarks.